standard/bridge-v1/pkg/relayer/relayer.go

```diff
@@ -26,94 +30,105 @@ type Options struct {
 }
 
 type Relayer struct {
+	logger *slog.Logger
 	// Closes ctx's Done channel and waits for all goroutines to close.
 	waitOnCloseRoutines func()
 	db                  *sql.DB
 }
 
-func NewRelayer(opts *Options) *Relayer {
-	r := &Relayer{}
+func NewRelayer(opts *Options) (*Relayer, error) {
+	r := &Relayer{logger: opts.Logger}
 
 	pubKey := &opts.PrivateKey.PublicKey
 	pubKeyBytes := crypto.FromECDSAPub(pubKey)
 	hash := sha3.NewLegacyKeccak256()
 	hash.Write(pubKeyBytes[1:])
 	address := hash.Sum(nil)[12:]
-	valAddr := common.BytesToAddress(address)
 
-	log.Info().Msg("Relayer signing address: " + valAddr.Hex())
+	r.logger.Info("relayer signing address", "address", common.BytesToAddress(address).Hex())
 
 	l1Client, err := ethclient.Dial(opts.L1RPCUrl)
 	if err != nil {
-		log.Fatal().Err(err).Msg("failed to dial l1 rpc")
+		return nil, fmt.Errorf("failed to dial l1 rpc: %w", err)
 	}
 
-	l1ChainID, err := l1Client.ChainID(context.Background())
+	l1ChainID, err := l1Client.ChainID(opts.Ctx)
 	if err != nil {
-		log.Fatal().Err(err).Msg("failed to get l1 chain id")
+		return nil, fmt.Errorf("failed to get l1 chain id: %w", err)
 	}
-	log.Info().Msg("L1 chain id: " + l1ChainID.String())
+	r.logger.Info("L1 chain id", "chain_id", l1ChainID)
 
 	settlementClient, err := ethclient.Dial(opts.SettlementRPCUrl)
 	if err != nil {
-		log.Fatal().Err(err).Msg("failed to dial settlement rpc")
+		return nil, fmt.Errorf("failed to dial settlement rpc: %w", err)
 	}
 
-	settlementChainID, err := settlementClient.ChainID(context.Background())
+	settlementChainID, err := settlementClient.ChainID(opts.Ctx)
 	if err != nil {
-		log.Fatal().Err(err).Msg("failed to dial settlement rpc")
+		return nil, fmt.Errorf("failed to dial settlement rpc: %w", err)
 	}
-	log.Info().Msg("Settlement chain id: " + settlementChainID.String())
-
-	ctx, cancel := context.WithCancel(context.Background())
+	r.logger.Info("settlement chain id", "chain_id", settlementChainID)
 
 	sFilterer, err := shared.NewSettlementFilterer(opts.SettlementContractAddr, settlementClient)
 	if err != nil {
-		log.Fatal().Err(err).Msg("failed to create settlement filterer")
+		return nil, fmt.Errorf("failed to create settlement filterer: %w", err)
 	}
-	sListener := NewListener(settlementClient, sFilterer, false)
+
+	ctx, cancel := context.WithCancel(opts.Ctx)
+	defer cancel()
+
+	sListener := NewListener(r.logger.With("component", "settlement_listener"), settlementClient, sFilterer, false)
 	sListenerClosed, settlementEventChan, err := sListener.Start(ctx)
 	if err != nil {
-		log.Fatal().Err(err).Msg("failed to start settlement listener")
+		return nil, fmt.Errorf("failed to start settlement listener: %w", err)
```

Comment:
One option to still remove the `defer cancel()` mentioned above, while avoiding compiler warnings is to add `cancel()` calls to each of these error blocks. Maybe there's a simpler option too